### ISSUE TITLE
After training Qwen3-MoE, the model's performance is very poor.

### ISSUE THREAD

**Issue by yanrui27** [snowflakedb/ArcticTraining#250]
When I was training the Qwen3 MoE series, the loss behaved normally. 

However, after training, the model was unable to consistently output tool calls, and its performance was very poor. 

 Using the same data and training configuration with Transformers + DeepSpeed, everything worked fine.

**Comment by yanrui27** [comment]
I observed that when training the MoE model, the first step usually takes 10 times longer.
For example:
0/100 — 1 minute
0/100 — 10 minutes
1/100 — 1 minute
....

**Comment by xylian86** [comment]
https://github.com/deepspeedai/DeepSpeed/pull/4966

**Comment by yanrui27** [comment]
https://github.com/snowflakedb/ArcticTraining/issues/250#issuecomment-3137515349
Thank you so much for your help. Your suggestion allowed me to successfully train the MoE model.

However, I've noticed that under the same 128-GPU setup, DeepSpeed runs at about 1/10th the speed of Megatron. Do you have any insights on this issue?

**Comment by yanrui27** [comment]
> [deepspeedai/DeepSpeed#4966](https://github.com/deepspeedai/DeepSpeed/pull/4966)
> 
> Leaf module is required for MoE model.

Thank you so much for your help. Your suggestion allowed me to successfully train the MoE model.

However, I've noticed that under the same 128-GPU setup, DeepSpeed runs at about 1/10th the speed of Megatron. Do you have any insights on this issue?

**Comment by xylian86** [comment]
> DeepSpeed runs at about 1/10th the speed of Megatron

Could you provide more context—such as the ZeRO stage used, and the overall parallelism strategy employed in Megatron?

**Comment by yanrui27** [comment]
> > DeepSpeed runs at about 1/10th the speed of Megatron
> 
> Could you provide more context—such as the ZeRO stage used, and the overall parallelism strategy employed in Megatron?


Okay, below are the training configurations for DeepSpeed and Megatron, respectively.

**DeepSpeed**
```
epochs: 3
micro_batch_size: 1
sequence_parallel_size: 8
gradient_accumulation_steps: 16
activation_checkpoint_cpu_offload: true

deepspeed:
  seq_parallel_communication_data_type: bf16
  zero_optimization:
    allgather_bucket_size: 200000000
    allgather_partitions: true
    contiguous_gradients: true
    offload_param:
      device: cpu
    overlap_comm: true
    reduce_bucket_size: 200000000
    reduce_scatter: true
    stage: 3
    stage3_max_live_parameters: 1000000000
    stage3_max_reuse_distance: 1000000000
    stage3_param_persistence_threshold: 100000
    stage3_prefetch_bucket_size: 200000000

model:
  attn_implementation: flash_attention_2
  name_or_path: Qwen3-235B-A22B-Instruct-2507
  save_name: checkpoints/Qwen3-235B-A22B-Instruct-2507
  type: liger
```


**megatron**
```
--tensor-model-parallel-size 4 \
--pipeline-model-parallel-size 8 \
--context-parallel-size 2 \
--sequence-parallel \
--expert-tensor-parallel-size 1 \
--expert-model-parallel-size 8 \
--moe-ffn-hidden-size 1536 \
--moe-router-load-balancing-type aux_loss \
--moe-aux-loss-coeff 0.0000001 \
--moe-layer-freq 1 \
--moe-router-pre-softmax \
--moe-router-dtype fp32 \
--moe-permute-fusion \
--moe-enable-deepep 
```

**Comment by yanrui27** [comment]
https://github.com/QwenLM/Qwen3/issues/736

Perhaps due to the nature of MoE, when I train dense models, the speed is at least slightly slower than Megatron, but there isn’t a several-fold difference.









询问 ChatGPT

**Comment by xylian86** [comment]
```
offload_param:
      device: cpu
```

Could you try running without offloading parameters to the CPU and measure the throughput?

**Comment by sfc-gh-truwase** [comment]
@sfc-gh-sbekman FYI

**Comment by sfc-gh-sbekman** [comment]
> @sfc-gh-sbekman FYI

AFAIK, we haven't tried MoE models yet - AT isn't optimized for that yet, but we will be looking into it some time soon.

> When I set sp=1, I was surprised to find that the training time dropped from 120 seconds to 70 seconds.

I'm not sure how one can compare 2 frameworks the way you did. There are many defaults and settings that can be quite different. So just saying framework A is slower than framework B unless everything is matched is not going to work.

`sp>1` will definitely add an overhead because of the additional comms which are exposed and if you're on a slow network that could explain the very large overhead.

Could you please share your Megatron config?

> Using the same data and training configuration with Transformers + DeepSpeed, everything worked fine.

Did you mean to say "HF Trainer" or "HF Accelerate" + Deepspeed? I'm not sure what "Transformers + DeepSpeed" implies.

**Comment by yanrui27** [comment]
> AFAIK, we haven't tried MoE models yet - AT isn't optimized for that yet, but we will be looking into it some time soon.据我所知，我们尚未尝试过 MoE 模型——AT 目前还未针对此类模型进行优化，但我们会尽快着手研究。
> ...

Thank you very much for your reply. Let me reorganize the timeline:

1. Initially, I found that AT training with MoE was extremely slow at the first step, usually taking 10 times longer than a normal step.
2. Without fixing the issue in (1), I updated the DeepSpeed version (from 0.17.0 to 0.17.3) and found that AT would completely hang. Training with the HF trainer also hung, which strongly indicated that the issue was with DeepSpeed.
3. **xylian86** suggested using the leaf module to handle the MoE layer, and I found this to be very effective.

---

## After this, MoE training was able to run normally.

4. Afterwards, I used AT to train MoE (Qwen3-235B-A22B with 128\*H800 GPUs).
5. From previous experiments training dense models with AT, I observed two things:

   1. AT training speed is roughly comparable to Megatron, with no significant performance gap.
   2. Training time remains roughly the same when `sp` is set to 1, 2, 4, or 8.
6. However, when training MoE (Qwen3-235B-A22B with 128\*H800 GPUs), I was surprised to find:

   * Using the AT framework with `sp=8`, the training time was significantly slower than Megatron (several times slower).
7. When training MoE (Qwen3-30B-A3B with 32\*H800 GPUs), I observed that:

   * With `sp=1`, the speed was roughly the same as Megatron (70s:70s).
   * With `sp=2`, the speed was significantly slower than Megatron (120s:70s).
8. Therefore, I suspect that when `sp > 1`, there are unexpected issues arising during MoE model training.

**Comment by yanrui27** [comment]
> `sp>1` will definitely add some overhead because of the additional comms which are exposed and if you're on a slow network that could explain some of the overhead. But I see you mentioned that you didn't see much of a difference with dense models.

1. When training dense models with AT, there is no significant difference in training time between sp=1 and sp=8.
2. However, when training MoE models with AT, I found that the training time with sp=2 is almost twice that of sp=1. (Using 32×H800 GPUs to train the 30B-A3B MoE model)

> and your hardware setup - how many gpus are you using?
I used 32×H800 GPUs to train the 30B-a3b MoE model, and 128×H800 GPUs to train the 235B-a22b MoE model.




> Did you mean to say "HF Trainer" or "HF Accelerate" + Deepspeed? I'm not sure what "Transformers + DeepSpeed" implies.
Yes, by "Transformers + DeepSpeed," I mean "HF Trainer."

**Comment by sfc-gh-sbekman** [comment]
> Without fixing the issue in (1), I updated the DeepSpeed version (from 0.17.0 to 0.17.3) and found that AT would completely hang. Training with the HF trainer also hung, which strongly indicated that the issue was with DeepSpeed.

I wonder if this is related to this PR of mine: https://github.com/deepspeedai/DeepSpeed/pull/7266
if you revert it does the problem go away?

As you can see from my comments different pytorch/nccl versions have this hanging issue - I thought that I was successful at conditioning on the right versions but perhaps I missed some variations. the bug is in pytorch when `device_id` is used.

which pytorch version are you using? the output of this please:
```
python -c 'import torch; print(f"pt={torch.__version__}, cuda={torch.version.cuda}, nccl={torch.cuda.nccl.version()}")'
```
